### PR TITLE
[WIP] Make Pooch lazily create directories for python 3.7+

### PR DESCRIPTION
**Pull request by hmaarrfk**
## Description

@jni just opening this up as a potential solution for #4660

## Checklist

- [ ] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [ ] Gallery example in `./doc/examples` (new features only)
- [ ] Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- [ ] Unit tests
- [ ] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

## For reviewers

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

### CODE REVIEW

File: skimage/__init__.py [hunk 1]
```diff
@@ -130,7 +130,17 @@ def _raise_build_error(e):
                              img_as_ubyte,
                              img_as_bool,
                              dtype_limits)
-    from .data import data_dir
     from .util.lookfor import lookfor
 
+    if sys.version_info < (3, 7):
+        from .data import data_dir
+
+def __getattr__(name):
```
Comment (emmanuelle) [inline skimage/__init__.py:138]: very nice use of py3.7+ lazy loading :-).

Comment (hmaarrfk) [reply skimage/__init__.py:138]: Yeah, lets work on this in a bit. I want to get this PR to be really tiny, since it is hard to test in a pytest way.

**Comment by jni** [comment]
@hmaarrfk thanks! Can you rebase now that #4666 is merged?

`lazy_pooch` is a great branch name, reminds me of my greyhound. =D This is him right now:

IMG_4310

😃

File: skimage/__init__.py (hunk 1 above)
Comment (jni) [reply skimage/__init__.py:138]: Honestly that was my quick and dirty attempt. I am convinced there is an elegant solution here without even global variables, but it's been a long week (gave a skimage+napari workshop today) and I'm just not seeing it. I'm happy to have another pass at this in a few days after recharging my batteries, but anyone should feel free to give it a go in the meantime!

Comment (hmaarrfk) [reply skimage/__init__.py:138]: I think image_fetcher needs to be made a class.... Was thinking about this problem in the shower and you know that is where the best ideas come!

**Comment by hmaarrfk** [comment]
ok i rebased, but i must admit it needs serious review in light of some organizational changes in the original branch. sorry about that one.

I'm also still not convinced that `exist_ok=True` doesn't fix things for the affected user in the original issue.

**Comment by emmanuelle** [comment]
@jni @hmaarrfk if you think this PR is going to take some time to get merged in I can also release 0.17.2 with pooch an optional dependency and 0.17.3 after this one is merged. Please tell me!

**Comment by jni** [comment]
@emmanuelle please do it! Optional pooch will already solve the problem for the OP of #4660 since they are not using the sample data at all.

File: skimage/__init__.py (hunk 1 above)
Comment (jni) [reply skimage/__init__.py:138]: You're right. The whole problem is that we have to keep track of a lot of state — that's what classes are for! I was stubbornly treating the whole module as a big class.

**Comment by hmaarrfk** [comment]
Definitely do it. This will help people building on automated machines for sure. Unless they use conda...

**Comment by sciunto** [comment]
THe issue #4660 is solved. Is this still useful in any sense?

**Comment by hmaarrfk** [comment]
I think the correct solution is a two pronged approach:

Leaving the `data_dir` pointing to the old datasets. 

Allowing users that have pooch installed use the newer, larger datasets.

I'm not super convinced, I have all the use cases right, I'll try to write a spec.

**Pull request closed without merging**